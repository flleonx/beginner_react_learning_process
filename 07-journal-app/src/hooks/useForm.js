import { useState } from "react";

export const useForm = ( initialState = {} ) => {

  const [values, setValues] = useState( initialState );

  const reset = ( newFormState = initialState ) => {
    setValues( newFormState );
    /*
      This could be redundant, although useState ignore its argument
      after the first render, if initialState change in the component
      it's passed with its current value, that is,
      setValues( newFormState ) == setValues( initialState )
    */
  };

  const handleInputChange = ({ target }) => {
    setValues({
      ...values,
      [ target.name ]: target.value
    });
  };

  return [ values, handleInputChange, reset ];

};
