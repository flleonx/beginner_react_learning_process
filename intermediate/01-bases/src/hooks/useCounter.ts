import { useState, useRef, useEffect, useLayoutEffect } from "react";
import { gsap } from 'gsap';

export const useCounter = ({ maxCount = 1 }) => {

  const [counter, setCounter] = useState(5);
  const counterHTMLElement = useRef<HTMLHeadingElement>(null);
  const timeLine = useRef(gsap.timeline());

  const handleClick = () => {
    // setCounter( prev => (prev >= MAXIMUN_COUNT) ? MAXIMUN_COUNT : prev + 1);
    setCounter(prev => Math.min(prev + 1, maxCount));
  };

  /*
    It's possible use useLayoutEffect when we strictly need that the effect
    to be executed after the HTML is rendered.
  */

  useLayoutEffect(() => {

    if(!counterHTMLElement.current) return;

    timeLine.current.to(counterHTMLElement.current, { y: -10, duration: 0.2, ease: "ease.out" })
      .to(counterHTMLElement.current, { y: 0, duration: 1, ease: "bounce.out" })
      .pause()

  }, []);

  /*
    React team suggest that each useEffect should
    have a single responsability
  */

  useEffect(() => {
    // if(counter < 10) return;

    console.log("%cMax value", "color: red; background-color: black;");

    // Reproduce the animation from the beginning when counter changes
    timeLine.current.play(0);

  }, [counter]);

  return {
    counter,
    elementToAnimate: counterHTMLElement,
    handleClick
  };
};
