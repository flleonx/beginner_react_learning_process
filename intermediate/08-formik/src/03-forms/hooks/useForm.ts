import { useState } from "react";

export const useForm = <T>(initState: T) => {

  const [formData, setFormData] = useState(initState);

  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    console.log(event.target.name);
    setFormData((prev) => ({
      ...prev,
      [event.target.name]: event.target.value,
    }));
  };

  const resetForm = () => {
    setFormData({ ...initState });
  };

  return {
    ...formData,
    // Properties
    formData,
    // Methods
    onChange,
    resetForm
  };
};
