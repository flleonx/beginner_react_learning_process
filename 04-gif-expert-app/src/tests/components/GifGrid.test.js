import React from "react";
import { shallow } from 'enzyme';
import GifGrid from '../../components/GifGrid';
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock('../../hooks/useFetchGifs');

describe('Test on <GifGrid />', () => {

  const category = 'Fate Zero'

  test('It should show <GifGrid /> properly', () => {

    // Here the values are simulated
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true
    });

    const wrapper = shallow( <GifGrid category={ category } /> );
    expect( wrapper ).toMatchSnapshot();

  });

  test('It should show items when the images are loaded', () => {

    const gifs = [{
      id: 'ABC',
      url: "https://localhost/anything/img.gif",
      title: "Anything"
    },
    {
      id: '123',
      url: "https://localhost/anything/img.gif",
      title: "Anything"
    }];

    // Here the values are simulated
    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false
    });

    const wrapper = shallow( <GifGrid category={ category } /> );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find('p').exists() ).toBe(false);
    expect( wrapper.find('GifGridItem').length ).toBe( gifs.length );

  });

});
