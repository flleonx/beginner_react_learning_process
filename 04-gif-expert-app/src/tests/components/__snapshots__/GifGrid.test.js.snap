// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <GifGrid /> It should show <GifGrid /> properly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Fate Zero
  </h3>
  <p
    className="animate__animated__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Test on <GifGrid /> It should show items when the images are loaded 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Fate Zero
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Anything"
      url="https://localhost/anything/img.gif"
    />
    <GifGridItem
      id="123"
      key="123"
      title="Anything"
      url="https://localhost/anything/img.gif"
    />
  </div>
</Fragment>
`;
